PFW-1530 Add "Tune" button to IDLER CANNOT HOME screen

gudnimg opened the pull request:
TODOs:
- [x] Change "Idler" text to something more descriptive while also being less than 17 characters
- [x] Add similar options to SELECTOR CANNOT HOME screen ( with range 1 to 4)
- [x] Try to change the screen such that it reads the current value from MMU EEPROM instead of showing a hard-coded value on the printer side.

----

Instead of adding a new Stallguard Threshold editing menu to the LCD Experimental menu, I propose to add a new _Tune_ button option to the MMU error screen. The Experimental menu is not accessible at all when the MMU error screen is active and so I think having the options available during the MMU error UI itself is more user friendly. Which ever way we go there is a lot of memory consumed by this since the menu code is very memory (flash) hungry.

Below are some images to show the new UI. Note that "Idler" text will be changed, I only took these photos to show the concept.

When the _Done_ button is clicked, the MMU error screen will disappear and the Idler will attempt to re-home with the new setting.

| Error screen with new option| Tune menu |
|------|------|
| ![IMG_3793](https://github.com/prusa3d/Prusa-Firmware/assets/8218499/f679601f-a3fb-496d-bf42-cf5c3aed0240)    |   ![IMG_3794](https://github.com/prusa3d/Prusa-Firmware/assets/8218499/f213e71f-9cb6-4d38-ac74-a22937e3ec08)   |

gudnimg committed:
PFW-1530 Initial prototype

Add a 'Tune' option to HOMING_IDLER_FAILED error

This will open a menu which allows
the user to change the stallguard threshold
from the MMU error screen

Change in memory:
Flash: +334 bytes
SRAM: +1 byte

gudnimg committed:
PFW-1530 Minor optimisation

Change in memory:
Flash: -8 bytes
SRAM: 0 bytes

gudnimg commented:
Synced the branch with MK3, hopefully the reported PR size is now more correct.

Ro3Deee commented:
I reduced the stall guard even if it was working ok with the default value,because it works smoother with a lower value. So i think it should be available always.
 We should also be able to able to have an option to "stop"/cancel the error message , than change the value and then restart the mmu.

gudnimg commented:
> We should also be able to able to have an option to "stop"/cancel the error message , than change the value and then restart the mmu.

I agree but currently the way we designed the protocol interface it's not possible. The MMU will report an error every 1 second (for every `Q0` Query). And the printer simply checks => if there is no error screen, then show it.

I'd also like to see some way to Cancel MMU operations, but we're not there yet (not highest priority).

gudnimg commented:
The point with this is to at least get the homing working so there is no error screen reported. Maybe we can re-use the same code in the Experimental menu without using more Flash. But as this is still a draft, we'll see. I have a lot of cleanup to do yet.

Ro3Deee commented:
Another option would be to be able to send configuration g codes from Prusa slicer and also configure the printer using a visual interface with menus and windows there, as in PS there is no memory limitation for menus. I didn't noticed that pronterface upcased x letter in hexadecimal register number (until you told me).

Sending the config g codes seems to work even if error message is active on lcd.

SNDR0 commented:
Idler ----> Engager, coupler, REF(Rotary Engager Filament)?

gudnimg committed:
mmu2: cleanup: use Register enum

gudnimg committed:
PFW-1530 Add Tune to Selector screen and more

- Rename "Idler" to "Sensitivity"
- Implement ReadRegisterInner() as a way to read register in blocking contexts such as manage_response()
This allows us to show the current EEPROM value on the printer's LCD

gudnimg committed:
PFW-1530 minor optimisation

Don't inline WriteRegister

Change in memory:
Flash: -12 bytes
SRAM: 0 bytes

gudnimg commented:
@DRracer I wonder if we can improve `ReadRegisterInner()` 🤔 It now adds ~100B

gudnimg committed:
PFW-1530 Refactor the menu to be non-blocking

The UI works much smoother now

Saves ~60 bytes of flash

gudnimg commented:
Opening PR for review... I have fixed all bugs I've found so far and tried to minimize flash usage, but this could use some more testing.

gudnimg committed:
PFW-1530 fixup

gudnimg commented:
Video of the menu currently. I trigger Selector Cannot Home by holding it in place with my hand. Hopefully Github doesnt compress the video too much 👀

https://github.com/prusa3d/Prusa-Firmware/assets/8218499/4e41fa78-9df1-4e32-9674-5401b4f8c102

gudnimg committed:
PFW-1530 cleanup: remove redundant uint8_t cast

DRracer commented on Firmware/mmu2.h:
```diff
@@ -297,6 +307,7 @@ class MMU2 {
     ErrorCode lastErrorCode = ErrorCode::MMU_NOT_RESPONDING;
     ErrorSource lastErrorSource = ErrorSource::ErrorSourceNone;
     Buttons lastButton = Buttons::NoButton;
+    uint16_t lastReadRegisterValue = 0;
```
Not important now, but may be make a pair<uint8_t register index, uint16_t value> - to make sure the last read value is related to a specific register index. This will become important when we start reading multiple registers on demand.

DRracer commented on Firmware/mmu2_reporting.cpp:
```diff
@@ -216,11 +218,22 @@ enum ReportErrorHookStates ReportErrorHookState = ReportErrorHookStates::RENDER_
     return is_mmu_error_monitor_active;
 }
 
+bool isErrorScreenSleeping() {
```
May be we could call it "TuneMenuEntered"

gudnimg replied:
Done

DRracer commented on Firmware/mmu2_reporting.cpp:
```diff
@@ -359,4 +372,68 @@ void ScreenClear(){
+static const TuneItem TuneItems[] PROGMEM = {
+  { (uint8_t)Register::Selector_sg_thrs_R, 1, 4},
+  { (uint8_t)Register::Idler_sg_thrs_R, 4, 7},
+};
```
:+1:

DRracer commented on Firmware/mmu2_reporting.cpp:
```diff
@@ -359,4 +372,68 @@ void ScreenClear(){
+
+static_assert(sizeof(TuneItems)/sizeof(TuneItem) == 2);
+
+typedef struct
```
let's make it a C++ struct

gudnimg replied:
Do I understand correctly you mean to drop `typedef`?

gudnimg replied:
Done

DRracer reviewed (commented):
Actually, the code is quite clean, I don't see any significant design flaw. I added a few minor comments.

gudnimg committed:
PFW-1530 Remove typedef keyword

gudnimg committed:
PFW-1530 rename isErrorScreenSleeping to TuneMenuEntered
